src/com/trilead/ssh2/crypto/dh/DhExchange.java

```diff
@@ -1,160 +1,212 @@
+			"12BF2D5B0B7474D6E694F91E6DBE115974A3926F12FEE5E4" +
+			"38777CB6A932DF8CD8BEC4D073B931BA3BC832B68D9DD300" +
+			"741FA7BF8AFC47ED2576F6936BA424663AAB639C5AE4F568" +
+			"3423B4742BF1C978238F16CBE39D652DE3FDB8BEFC848AD9" +
+			"22222E04A4037C0713EB57A81A23F0C73473FC646CEA306B" +
+			"4BCBC8862F8385DDFA9D4B7FA2C087E879683303ED5BDD3A" +
+			"062B3CF5B3A278A66D2A13F83F44F82DDF310EE074AB6A36" +
+			"4597E899A0255DC164F31CC50846851DF9AB48195DED7EA1" +
+			"B1D510BD7EE74D73FAF36BC31ECFA268359046F4EB879F92" +
+			"4009438B481C6CD7889A002ED5EE382BC9190DA6FC026E47" +
+			"9558E4475677E9AA9E3050E2765694DFC81F56E880B96E71" +
+			"60C980DD98EDD3DFFFFFFFFFFFFFFFFF", 16);
+
+	private static final BigInteger G = BigInteger.valueOf(2);
+
+	/* Hash algorithm to use */
+	private String hashAlgo;
 
 	/* Client public and private */
 
-	BigInteger e;
-	BigInteger x;
+	private DHPrivateKey clientPrivate;
+	private DHPublicKey clientPublic;
 
 	/* Server public */
 
-	BigInteger f;
-
-	/* Shared secret */
-
-	BigInteger k;
-
-	static
-	{
-		final String p1_string = "17976931348623159077083915679378745319786029604875"
-				+ "60117064444236841971802161585193689478337958649255415021805654859805036464"
-				+ "40548199239100050792877003355816639229553136239076508735759914822574862575"
-				+ "00742530207744771258955095793777842444242661733472762929938766870920560605"
-				+ "0270810842907692932019128194467627007";
-
-		final String p14_string = "FFFFFFFFFFFFFFFFC90FDAA22168C234C4C6628B80DC1CD129"
-				+ "024E088A67CC74020BBEA63B139B22514A08798E3404DDEF9519B3CD3A431B302B0"
-				+ "A6DF25F14374FE1356D6D51C245E485B576625E7EC6F44C42E9A637ED6B0BFF5CB"
-				+ "6F406B7EDEE386BFB5A899FA5AE9F24117C4B1FE649286651ECE45B3DC2007CB8A"
-				+ "163BF0598DA48361C55D39A69163FA8FD24CF5F83655D23DCA3AD961C62F356208"
-				+ "552BB9ED529077096966D670C354E4ABC9804F1746C08CA18217C32905E462E36C"
-				+ "E3BE39E772C180E86039B2783A2EC07A28FB5C55DF06F4C52C9DE2BCBF69558171"
-				+ "83995497CEA956AE515D2261898FA051015728E5A8AACAA68FFFFFFFFFFFFFFFF";
-
-		p1 = new BigInteger(p1_string);
-		p14 = new BigInteger(p14_string, 16);
-		g = new BigInteger("2");
-	}
-
-	private final String hashAlgorithm;
-
-	@Deprecated
-	public DhExchange()
-	{
-		this("SHA1");
-	}
-
-	public DhExchange(String hashAlgorightm) {
-		super();
-		this.hashAlgorithm = hashAlgorightm;
-	}
-
-	public void init(int group, SecureRandom rnd)
-	{
-		k = null;
-
-		if (group == 1)
-			p = p1;
-		else if (group == 14)
-			p = p14;
-		else
-			throw new IllegalArgumentException("Unknown DH group " + group);
-
-		x = new BigInteger(p.bitLength() - 1, rnd);
-
-		e = g.modPow(x, p);
-	}
-
-	/**
-	 * @return Returns the e.
-	 *
-	 */
-	public BigInteger getE()
-	{
-		if (e == null)
-			throw new IllegalStateException("DhDsaExchange not initialized!");
+	private DHPublicKey serverPublic;
+
+	@Override
+	public void init(String name) throws IOException {
+		final DHParameterSpec spec;
+		if ("diffie-hellman-group18-sha512".equals(name)) {
```

Comment:
Nit: this can use a `switch (name)` statement.

Reply:
Done!